### PR TITLE
Move yk-internal parts from `ykcapi` to `ykrt`

### PR DESCRIPTION
**Pull request by ltratt**
`ykcapi` is a good name for what we want that crate to do, but the actual contents of `ykcapi` have become "a C api and random internal yk code that needs to be part of a cdylib".

This PR moves nearly all of that internal yk code from `ykcapi` to `ykrt`. The only thing that's left is `__ykrt_control_point` -- and, arguably, there isn't a much better place for that than `ykcapi` (or, at least, I can't think of a much better place for it).

### CODE REVIEW

**Comment by vext01** [comment]
Earlier on we had the rule "if possible, put all C functions in ykcapi, to avoid linkage issues".

I no longer remember what those issues were...

File: ykrt/src/deopt.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
+//! Run-time deoptimisation support: when a guard fails, this module restores the state necessary
+//! to resume interpreter execution.
+
+use std::{arch::asm, ffi::c_void};
+
+/// The `__llvm__deoptimize()` function required by `llvm.experimental.deoptimize` intrinsic, that
+/// we use for exiting to the stop-gap interpreter on guard failure.
```
Comment (vext01) [inline ykrt/src/deopt.rs:87]: mention of stop-gap interpreter here.

**Comment by vext01** [comment]
If it links, I'm OK with it.

How do we feel about the `__<crate>_<thing>` convention used? I often question it. I guess it doesn't hugely matter.

**Comment by ltratt** [comment]
> Earlier on we had the rule "if possible, put all C functions in ykcapi, to avoid linkage issues".

I did encounter some linking issues when doing this PR, not all of which I understood. I *think* in retrospect, they were because I was splitting some things across two different crates when they needed to be in one crate.

File: ykrt/src/deopt.rs (hunk 1 above)
Comment (ltratt) [reply ykrt/src/deopt.rs:87]: There are quite a lot of "stop-gap"s in the repo, including in test file names. If we like this PR, I will do another PR which goes through and fixes all those (I worried that doing so here would make the PR hard to read).

Comment (vext01) [reply ykrt/src/deopt.rs:87]: sounds good!

**Comment by vext01** [comment]
Okie dokie!

bors r+

**Comment by ltratt** [comment]
bors r-

I would like to check @ptersilie is OK with this too before merging.

**Comment by vext01** [comment]
My apologies

**Comment by ptersilie** [comment]
Looks good to me.

**Pull request merged**